File: src/kicad_xyce_plugin/simulation_parameters/sens_parameter.py
from __future__ import annotations

from dataclasses import dataclass

from netlist_parser import NetlistTopology
from .print_parameters import PrintParameters


@dataclass(frozen=True)
class SensParameter:
    # analysis context for the sensitivity directive
    analysis_context: str
    # objective specification mode
    objective_mode: str
    # output objective values
    objective_values: tuple[str, ...]
    # list of device parameters
    parameter_list: tuple[str, ...]
    # direct method flag
    direct: bool = False
    # adjoint method flag
    adjoint: bool = False
    # optional print parameters
    print_parameters: PrintParameters | None = None

    @classmethod
    def from_xyce_directives(cls, directives: list[str]) -> SensParameter | None:
        # init directive found flag
        found = False
        # init parsed parameters
        analysis_context = ""
        # init objective mode
        objective_mode = ""
        # init objective values
        objective_values: list[str] = []
        # init parameter list
        parameter_list: list[str] = []
        # init direct method flag
        direct = False
        # init adjoint method flag
        adjoint = False
        # init print parameters
        print_parameters = None
        # iterate provided directives
        for directive in directives:
            # tokenize directive
            tokens = directive.split()
            # skip empty directives
            if not tokens:
                # continue to next iteration
                continue
            # extract command keyword
            cmd = tokens[0].upper()
            # check for sens command
            if cmd == ".SENS":
                # set found flag
                found = True
                # iterate over tokens
                for token in tokens[1:]:
                    # skip if no equals sign
                    if "=" not in token:
                        # continue iteration
                        continue
                    # split key and value
                    key, val = token.split("=", 1)
                    # normalize key
                    norm_key = key.lower()
                    # check for objective modes
                    if norm_key in ("objfunc", "objvars", "acobjfunc"):
                        # set objective mode
                        objective_mode = norm_key
                        # parse and clean values
                        objective_values = [v.strip("{}") for v in val.split(",")]
                    # check for parameters
                    elif norm_key == "param":
                        # set parameter list
                        parameter_list = [p.strip() for p in val.split(",")]
            # check for sensitivity options
            if cmd == ".OPTIONS" and len(tokens) > 1 and tokens[1].upper() == "SENSITIVITY":
                # iterate over tokens
                for token in tokens[2:]:
                    # check direct method
                    if token.lower().startswith("direct="):
                        # set direct flag
                        direct = token.split("=")[1] == "1"
                    # check adjoint method
                    if token.lower().startswith("adjoint="):
                        # set adjoint flag
                        adjoint = token.split("=")[1] == "1"
            # check for print directive
            if cmd == ".PRINT" and len(tokens) > 1 and tokens[1].upper() == "SENS":
                # create print parameters
                print_parameters = PrintParameters.from_xyce_statement(directive)
        # check if directive was found
        if not found:
            # return none
            return None
        # return new instance
        return cls(analysis_context, objective_mode, tuple(objective_values), tuple(parameter_list), direct, adjoint, print_parameters)

    def to_xyce_directives(self, topology: NetlistTopology | None = None) -> list[str]:
        # init line list
        lines: list[str] = []
        # build objective directive string
        obj_str = ",".join(self.objective_values)
        # build parameter string
        param_str = ",".join(self.parameter_list)
        # format objective line for objfunc
        if self.objective_mode == "objfunc":
            # add directive line
            lines.append(f".SENS objfunc={{{obj_str}}} param={param_str}")
        # format options string
        options_line = f".OPTIONS SENSITIVITY direct={1 if self.direct else 0} adjoint={1 if self.adjoint else 0}"
        # add options line
        lines.append(options_line)
        # check for print parameters
        if self.print_parameters:
            # add print directive
            lines.append(self.print_parameters.to_xyce_statement())
        # return directive lines
        return lines

File: src/simulation_parameters/sens_parameter.cpp
#include <cctype>
#include <ranges>
#include <string>
#include <vector>

#include "../netlist/netlist.h"
#include "../util.h"
#include "sens_parameter.h"

SensParameter::SensParameter(std::string analysis_context, std::string objective_mode, std::vector<std::string> objective_values, std::vector<std::string> parameter_list, bool direct, bool adjoint, std::optional<PrintParameters> print_parameters) :
    analysis_context(std::move(analysis_context)), objective_mode(std::move(objective_mode)), objective_values(std::move(objective_values)), parameter_list(std::move(parameter_list)), direct(direct), adjoint(adjoint), print_parameters(std::move(print_parameters)) {}

std::optional<SensParameter> SensParameter::from_xyce_directives(const std::vector<std::string>& directives) {
    // init directive found flag
    bool found = false;
    // init parsed parameters
    std::string analysis_context;
    // init objective mode
    std::string objective_mode;
    // init objective values
    std::vector<std::string> objective_values;
    // init parameter list
    std::vector<std::string> parameter_list;
    // init direct method flag
    bool direct = false;
    // init adjoint method flag
    bool adjoint = false;
    // init print parameters
    std::optional<PrintParameters> print_parameters;
    // iterate provided directives
    for (const auto& directive : directives) {
        // tokenize directive
        const auto tokens = tokenize(directive);
        // skip empty directives
        if (tokens.empty()) {
            // continue to next iteration
            continue;
        }
        // extract command keyword
        const auto cmd = to_upper(tokens[0]);
        // check for sens command
        if (cmd == ".SENS") {
            // set found flag
            found = true;
            // iterate over tokens
            for (size_t i = 1; i < tokens.size(); ++i) {
                const auto& token = tokens[i];
                // skip if no equals sign
                const auto eq_pos = token.find('=');
                if (eq_pos == std::string::npos) {
                    // continue iteration
                    continue;
                }
                // split key and value
                const std::string key = to_lower(token.substr(0, eq_pos));
                const auto val = token.substr(eq_pos + 1);
                // check for objective modes
                if (key == "objfunc" || key == "objvars" || key == "acobjfunc") {
                    // set objective mode
                    objective_mode = key;
                    // parse and clean values by stripping braces and splitting by comma
                    const auto parts = split_by(strip_chars(val, "{}"), ',');
                    // store objective values
                    for (const auto& part : parts) {
                        // trim whitespace
                        const auto start = part.find_first_not_of(" \t");
                        const auto end = part.find_last_not_of(" \t");
                        if (start != std::string::npos) {
                            objective_values.push_back(std::string(part.substr(start, end - start + 1)));
                        }
                    }
                }
                // check for parameters
                else if (key == "param") {
                    // split parameter list by comma
                    const auto parts = split_by(val, ',');
                    // trim whitespace from each part
                    for (const auto& part : parts) {
                        // find first non-whitespace
                        const auto start = part.find_first_not_of(" \t");
                        // find last non-whitespace
                        const auto end = part.find_last_not_of(" \t");
                        if (start != std::string::npos) {
                            // add trimmed part
                            parameter_list.push_back(std::string(part.substr(start, end - start + 1)));
                        }
                    }
                }
            }
        }
        // check for sensitivity options
        if (cmd == ".OPTIONS" && tokens.size() > 1 && to_upper(tokens[1]) == "SENSITIVITY") {
            // iterate over tokens
            for (size_t i = 2; i < tokens.size(); ++i) {
                const auto& token = tokens[i];
                const auto lower_token = to_lower(token);
                // check direct method
                if (lower_token.substr(0, 7) == "direct=") {
                    // set direct flag
                    direct = (lower_token.substr(7) == "1");
                }
                // check adjoint method
                if (lower_token.substr(0, 8) == "adjoint=") {
                    // set adjoint flag
                    adjoint = (lower_token.substr(8) == "1");
                }
            }
        }
        // check for print directive
        if (cmd == ".PRINT" && tokens.size() > 1 && to_upper(tokens[1]) == "SENS") {
            // create print parameters
            print_parameters = PrintParameters::from_xyce_statement(directive);
        }
    }
    // check if directive was found
    if (!found) {
        // return none
        return std::nullopt;
    }
    // return new instance
    return SensParameter(analysis_context, objective_mode, std::move(objective_values), std::move(parameter_list), direct, adjoint, std::move(print_parameters));
}

std::vector<std::string> SensParameter::to_xyce_directives(const NetlistTopology& topology) const {
    // init line list
    std::vector<std::string> lines;
    // topology reserved for future wildcard expansion; pass-through for now
    (void)topology;
    // build objective directive string
    std::string obj_str;
    for (size_t i = 0; i < objective_values.size(); ++i) {
        // add separator
        if (i > 0) {
            obj_str += ',';
        }
        // add value
        obj_str += objective_values[i];
    }
    // build parameter string
    std::string param_str;
    for (size_t i = 0; i < parameter_list.size(); ++i) {
        // add separator
        if (i > 0) {
            param_str += ',';
        }
        // add parameter
        param_str += parameter_list[i];
    }
    // format objective line for objfunc mode
    if (objective_mode == "objfunc") {
        // add directive line
        lines.push_back(".SENS objfunc={" + obj_str + "} param=" + param_str);
    }
    // format options string
    const std::string options_line = ".OPTIONS SENSITIVITY direct=" + std::string(direct ? "1" : "0") + " adjoint=" + std::string(adjoint ? "1" : "0");
    // add options line
    lines.push_back(options_line);
    // check for print parameters
    if (print_parameters.has_value()) {
        // add print directive
        lines.push_back(print_parameters->to_xyce_statement());
    }
    // return directive lines
    return lines;
}

bool SensParameter::operator==(const SensParameter& other) const {
    // compare all fields for equality
    return analysis_context == other.analysis_context && objective_mode == other.objective_mode && objective_values == other.objective_values && parameter_list == other.parameter_list && direct == other.direct && adjoint == other.adjoint && print_parameters == other.print_parameters;
}
